### PR TITLE
add dataset generation

### PR DESCRIPTION
**Pull request by waltaskew**
Implement Dataset Generator

builds off of https://github.com/UrbanSystemsLab/climateiq-cnn/pull/112

@Katsutoshii would appreciate some help with some better documentation

### CODE REVIEW

**Review by prashantkul — APPROVED** [review]

File: usl_models/usl_models/flood_ml/dataset.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+    logging.info("Retrieving temporal features from %s.", gcs_url)
+    temporal_array = _download_as_array(storage_client, gcs_url)
+    temporal_tensor = tf.convert_to_tensor(
+        numpy.tile(temporal_array, (m_rainfall, 1)).T,
+        dtype=tf.float32,
+    )
+    return temporal_tensor
+
+
+def _iter_feature_label_tensors(
```
Comment (Katsutoshii) [inline usl_models/usl_models/flood_ml/dataset.py:193]: Perhaps it's more clear to replace "feature" with "geo_feature" in this function and its callsite.

File: usl_models/usl_models/flood_ml/dataset.py [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+) -> tf.Tensor:
+    """Creates a temporal tensor from the numpy array stored in GCS."""
+    temporal_metadata = metastore.get_temporal_feature_metadata(
+        firestore_client, sim_name
+    )
+    gcs_url = temporal_metadata["as_vector_gcs_uri"]
+
+    logging.info("Retrieving temporal features from %s.", gcs_url)
+    temporal_array = _download_as_array(storage_client, gcs_url)
+    temporal_tensor = tf.convert_to_tensor(
```
Comment (Katsutoshii) [inline usl_models/usl_models/flood_ml/dataset.py:186]: nit: it's likely more efficient to download as tensor and then compute its tile + transpose. But given this is not a hot code path it's not critical.

File: usl_models/usl_models/flood_ml/dataset.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+    for t in range(T_max):
+        window_input = model.Input(
+            geospatial=model_input["geospatial"],
+            temporal=_extract_temporal(t, n_flood_maps, model_input["temporal"]),
+            spatiotemporal=_extract_spatiotemporal(t, n_flood_maps, labels),
+        )
+        yield window_input, labels[t]
+
+
+def _extract_temporal(t: int, n: int, temporal: tf.Tensor) -> tf.Tensor:
```
Comment (Katsutoshii) [inline usl_models/usl_models/flood_ml/dataset.py:112]: Let's add a similar docstring to _extract_temporal here.

File: usl_models/usl_models/flood_ml/dataset.py [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+
+    The examples are generated from multiple simulations.
+    They are windowed for training on next-map prediction.
+    The dataset iteratively yields examples read from Google Cloud Storage to avoid
+    pulling all examples into memory at once.
+
+    Args:
+      sim_names: The simulation labels to use for training,
+                 e.g. ["Manhattan-config_v1/Rainfall_Data_1.txt"]
+      batch_size: Size of batches yielded by the dataset.
```
Comment (Katsutoshii) [inline usl_models/usl_models/flood_ml/dataset.py:36]: Let's leave some memory hints here or point it at a doc that has more info on how to select batch size. E.g. "approx VRAM requirement = 10GB * batch_size during training".

File: usl_models/usl_models/flood_ml/dataset.py [hunk 5]
```diff
@@ -0,0 +1,231 @@
+import logging
```
Comment (Katsutoshii) [inline usl_models/usl_models/flood_ml/dataset.py:1]: Missing module docstring. Could be:
"""tf.data.Datasets for training FloodML model on CityCAT data."""

**Review by Katsutoshii — APPROVED** [review]
Overall LG. Left some minor comments and doc suggestions.

File: usl_models/usl_models/flood_ml/dataset.py (hunk 2 above)
Comment (waltaskew) [reply usl_models/usl_models/flood_ml/dataset.py]: now tiling on GPUs :)

**Pull request merged**